### PR TITLE
feat: cancellable api calls

### PR DESCRIPTION
**Pull request by achingbrain**
Passes a [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) to every API call which emits a `abort` event when the caller is no longer interested in the result of the operation.

Lower level code that creates resources or has other long-term effects should tear down those resources early if the `abort` event is received.

Adds support for `timeout` options to every API call that will emit an `abort` event on the passed signal and throw a Timeout error.

Finally `abort` events are triggered if the current request arrived via the HTTP API and the request was aborted from the client - that is, a `disconnect` event is fired by Hapi.

Updates the core-api docs to add these new options.

### CODE REVIEW

File: packages/ipfs-http-client/src/add.js [hunk 1]
```diff
@@ -5,21 +5,27 @@ const toCamel = require('./lib/object-to-camel')
...
         'stream-channels': true,
         ...options,
         progress: Boolean(progressFn)
       }),
       timeout: options.timeout,
-      signal: options.signal,
+      signal,
       ...(
-        await multipartRequest(input)
+        await multipartRequest(input, controller)
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply packages/ipfs-http-client/src/add.js:28]: found my answer below

File: packages/ipfs-http-client/src/lib/multipart-request.js [hunk 2]
```diff
@@ -49,6 +49,9 @@ async function multipartRequest (source, boundary = `---------------------------
 
         index++
       }
+    } catch (err) {
+      // workaround for https://github.com/node-fetch/node-fetch/issues/753
+      abortController.abort(err)
```
Comment (hugomrdias) [inline packages/ipfs-http-client/src/lib/multipart-request.js:54]: the change you made to `http` makes this error surface when we do?
```js
try {
api.post()
}catch (err){
	//here?
}

```
can't we pass the http instance instead of a new abortController and abort with the internal controller?

Comment (hugomrdias) [reply packages/ipfs-http-client/src/lib/multipart-request.js:54]: this would avoid having to do 
```
// allow aborting requests on body errors
    const controller = new AbortController()
    const signal = anySignal([controller.signal, options.signal]
````
everywhere

File: packages/ipfs/docs/MODULE.md [hunk 3]
```diff
@@ -21,7 +21,6 @@ Use the IPFS module as a dependency of your project to spawn in process instance
   - [`options.libp2p`](#optionslibp2p)
   - [Instance methods](#instance-methods)
     - [`node.start()`](#nodestart)
-    - [`node.stop()`](#nodestop)
```
Comment (hugomrdias) [inline packages/ipfs/docs/MODULE.md:24]: why did you remove stop ?

File: docs/core-api/BITSWAP.md [hunk 4]
```diff
@@ -29,15 +55,65 @@ console.log(list2)
...
+
+### Example
+
+```JavaScript
+const list = await ipfs.bitswap.wantlist()
+console.log(list)
+// [ CID('QmHash') ]
 
-##### `ipfs.bitswap.stat()`
+const list2 = await ipfs.bitswap.wantlist(peerId)
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply docs/core-api/BITSWAP.md:90]: Yeah, a lot of the docs have these copy/paste errors.  I tried to fix them where I noticed them..

File: packages/ipfs/docs/MODULE.md (hunk 3 above)
Comment (achingbrain) [reply packages/ipfs/docs/MODULE.md:24]: The README only has the additions to the core api - `.stop()` is documented [here](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/MISCELLANEOUS.md#stop) so this link was broken.

File: docs/core-api/BOOTSTRAP.md [hunk 5]
```diff
@@ -1,27 +1,50 @@
...
-  which to learn about other peers in the network.
+> Manipulates the bootstrap list, which contains the addresses of the bootstrap nodes. These are the trusted peers from which to learn about other peers in the network.
 
-> Only edit this list if you understand the risks of adding or removing nodes
+Warning: your node requires bootstrappers to join the network and find other peers.
 
-* [bootstrap.add](#bootstrapadd)
-* [bootstrap.list](#bootstraplist)
-* [bootstrap.rm](#bootstraprm)
+If you edit this list, you may find you have reduced or no connectivity.  If this is the case, please reset your node's bootstrapper list with `ipfs.bootstrap.add({ default: true })`.
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply docs/core-api/BOOTSTRAP.md:7]: inorite

File: packages/ipfs-http-client/src/lib/multipart-request.js (hunk 2 above)
Comment (achingbrain) [reply packages/ipfs-http-client/src/lib/multipart-request.js:54]: I'm not sure I follow, could you explain a bit more?

We need to create a new abort controller in `ipfs-utils/src/http` to support timeouts, and we can't reuse abort controller instances (once their signal is aborted, it's aborted forever) so using it in the next request makes the request abort immediately.

File: packages/interface-ipfs-core/src/files/ls.js [hunk 6]
```diff
@@ -196,5 +198,14 @@ module.exports = (common, options) => {
...
       expect(files.filter(file => file.name === fileName)).to.be.ok()
     })
+
+    it('should respect timeout option when listing files', async () => {
+      await testTimeout(() => drain(ipfs.files.ls({
+        timeout: 1
+      })))
+
+      // ensures that the request that timed out has completed
+      await ipfs.files.stat('/')
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/interface-ipfs-core/src/files/ls.js]: Hmm, I think it was necessary as `ipfsd-ctl` was failing to shut down nodes after the test run because the request was still in flight so datastores were open, etc but it shouldn't be - I've removed this everywhere.

File: packages/ipfs-http-client/src/lib/multipart-request.js (hunk 2 above)
Comment (hugomrdias) [reply packages/ipfs-http-client/src/lib/multipart-request.js:54]: in this PR https://github.com/ipfs/js-ipfs-utils/pull/41/files each request has it's own abort controller, so with that and changing multipart-request to receive the http instance we can call the internal abort controller, `http.abortController.abort()`
makes sense?

File: packages/ipfs/src/http/api/resources/files-regular.js [hunk 7]
```diff
@@ -294,19 +422,25 @@ exports.ls = {
...
-        const links = await all(ipfs.ls(key, {
-          recursive
+        const links = await all(ipfs.ls(path, {
+          recursive,
+          signal,
+          timeout
         }))
 
-        return h.response({ Objects: [{ Hash: key, Links: links.map(mapLink) }] })
+        return h.response({ Objects: [{ Hash: path.split('/')[2], Links: links.map(mapLink) }] })
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs/src/http/api/resources/files-regular.js:431]: My mistake.  I though `Hash` would be a hash, it's not, it's a path.

File: packages/ipfs/src/core/api-manager.js [hunk 8]
```diff
@@ -12,7 +12,7 @@ module.exports = class ApiManager {
     })
   }
 
-  update (nextApi, onUndef) {
+  update (nextApi, onUndef, noWrap) {
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs/src/core/api-manager.js:15]: It's not used, I've removed it.

File: packages/ipfs/src/http/api/resources/config.js [hunk 9]
```diff
@@ -202,9 +357,44 @@ exports.profiles = {
...
+          stripUnknown: true
+        },
+        query: Joi.object().keys({
+          timeout: Joi.timeout()
+        })
+          .rename('dry-run', 'dryRun', {
+            override: true,
+            ignoreUndefined: true
+          })
+          .rename('arg', 'profile', {
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs/src/http/api/resources/config.js:373]: I've removed it, and the dryRun arg too.

File: packages/ipfs/src/http/api/resources/files-regular.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs/src/http/api/resources/files-regular.js]: I've added a test for `ipfs.ls` with deep paths, I don't think we had one before.
